T2D-27 First prototype implementation

poggio84 opened the pull request:
- Implemented and tested the basic CRUD operations
- Ability to POST a file that will be archived (file system) and stored (tracked) 

T2D-27

Commit:
First file tracking prototype

Commit:
Update travis configuration

Commit:
Basic CRUD operations supported

Commit:
Fix for broken test

Commit:
Added jackson-datatype-jsr310 dependency to format Json Data

Commit:
Test the SourceFilePath persistence

Commit:
Id and hash are now separate in File

Commit:
File class renamed to FileMetadata

cyenyxe commented on pom.xml:
```diff
@@ -0,0 +1,118 @@
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.0.0.BUILD-SNAPSHOT</version>
```
Spring Boot 2.0 is scheduled for a November release. Can we stick to 1.5 for now?

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/FileManagerApplication.java:
```diff
@@ -0,0 +1,41 @@
+@EnableConfigurationProperties(StorageProperties.class)
+public class FileManagerApplication {
+
+	public static void main(String[] args) {
```
Fix indentation.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/FileType.java:
```diff
@@ -0,0 +1,31 @@
+ */
+package uk.ac.ebi.ampt2d;
+
+public enum FileType {
```
I guess that for consistency we should drop the `_FILE`.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/controller/FileTrackingController.java:
```diff
@@ -0,0 +1,76 @@
+    @Autowired
+    private FileRepository fileRepository;
+
+    @PostMapping("/upload-and-archive")
```
I would suggest using just `/upload` so we don't imply any submission-based process.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/controller/FileTrackingController.java:
```diff
@@ -0,0 +1,76 @@
+
+    @PostMapping("/upload-and-archive")
+    public ResponseEntity<?> handleFileUpload(@RequestParam("file") MultipartFile multipartFile) {
+
```
Remove blank line.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/controller/FileTrackingController.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            storedFile = convertMultipartToFile(multipartFile);
+        } catch (IOException e) {
+            e.printStackTrace();
```
If there is an issue the exception should be reported to the user. Can this be done via a `ResponseEntity` object with an HTTP error status?

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/controller/FileTrackingController.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private java.io.File convertMultipartToFile(MultipartFile file) throws IOException {
+        java.io.File convFile = new java.io.File(file.getOriginalFilename());
```
No need for the absolute package as our `File` class is now `FileMetadata`.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,162 @@
+        try {
+            this.hash = Files.hash(file, Hashing.sha384()).toString();
+        } catch (IOException e) {
+            e.printStackTrace();
```
This exception should be raised as the object won't be correct after construction if the hash is null.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,162 @@
+    public FileMetadata(String hash,
+                        String name,
+                        FileType type,
+                        long size) {
```
Fix format, all the parameters should fit in one line.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    public void setName(String name) {
+        Assert.hasText(hash, "FileMetadata name is required");
```
I think in this case the message should not refer to `FileMetadata` and the field checked is not correct.

```java
Assert.hasText(name, "File name is required");
```

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+
+    @ManyToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "file_id")
+    private FileMetadata fileMetadata;
```
Could the declaration of entity attributes be consistent across classes? I like the `FileMetadata` structure:

1. Class attributes
2. External references
3. Dates for auditing

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/repository/FileRepositoryCustom.java:
```diff
@@ -0,0 +1,24 @@
+import java.nio.file.Path;
+
+public interface FileRepositoryCustom {
+    Path archive(MultipartFile file);
```
`upload` would be more applicable to a multipart file.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/repository/FileRepository.java:
```diff
@@ -0,0 +1,27 @@
+import uk.ac.ebi.ampt2d.persistence.entities.FileMetadata;
+
+@RepositoryRestResource(collectionResourceRel = "files", path = "files")
+public interface FileRepository extends PagingAndSortingRepository<FileMetadata, String>, FileRepositoryCustom {
```
Fix formatting, method declarations should be separate by a blank line. Also applies to `FileRepositoryCustom`.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/storage/FileSystemStorageService.java:
```diff
@@ -0,0 +1,82 @@
+
+        try {
+            if (file.isEmpty()) {
+                throw new StorageException("Failed to store empty file " + file.getOriginalFilename());
```
The process doesn't really "fail", but it is not permitted. The message should reflect that.

Does this conditional branch need to be inside the try-catch? The [`isEmpty` method Javadoc](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/multipart/MultipartFile.html#isEmpty--) says nothing about an IOException being thrown.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/storage/FileSystemStorageService.java:
```diff
@@ -0,0 +1,82 @@
+                return resource;
+            }
+            else {
+                throw new StorageFileNotFoundException("Could not read file: " + filename);
```
Why not just using `FileNotFoundException`?

poggio84 replied:
Probably just to differentiate from the java.io one. The all folder is from [https://github.com/spring-guides/gs-uploading-files/tree/master/complete/src/main/java/hello/storage](url)

cyenyxe commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,32 @@
+# Naming strategy
+spring.jpa.hibernate.naming-strategy = org.hibernate.cfg.ImprovedNamingStrategy
+## The SQL dialect makes Hibernate generate better SQL for the chosen database
+spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.MySQL5Dialect
```
Could this be made compatible with PostgreSQL? That is the production database so it makes sense to use the same in development.

cyenyxe commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,32 @@
+server.port: 8081
+server.contextPath: /test
+
+
```
Add a comment about these 2 configuration parameters.

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/persistence/repository/FileMetadataRepositoryTest.java:
```diff
@@ -0,0 +1,120 @@
+
+    @Before
+    public void setUp() throws Exception {
+        fileRepository.deleteAll();
```
This must be unnecessary. If we need it then something is wrongly configured.

poggio84 replied:
The .`init` is creating the root dir where the files will be stored during the upload. The .`deleteAll` is removing the dir and the content to make sure it is starting from fresh. In production we will need to define the path in `StorageProperties` class and remove the init Bean

cyenyxe replied:
Do you mean the current implementation only works for testing? If so please modify it to make it work both for testing and production. We can discuss if necessary.

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/persistence/repository/FileMetadataRepositoryTest.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @Test
+    public void testCRUD() {
```
Be consistent when naming tests, either they start with "test" or they don't (the convention is to do so).

cyenyxe reviewed (changes requested):
Several files need code auto-formatting applied.

Commit:
FileType renamed to Type

Commit:
Fix indentation

Commit:
Improvements based on code review

Commit:
Removed init step

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+    private String hash;
+
+    @Enumerated(EnumType.STRING)
+    private Type type;
```
Are we gonna support null for type? If we don't know the file type shouldn't be OTHER?

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        return hash;
+    }
+
+    public void setHash(String hash) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        this.hash = hash;
+    }
+
+    public Type getType() {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        return type;
+    }
+
+    public void setType(Type type) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        this.type = type;
+    }
+
+    public long getSize() {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        return size;
+    }
+
+    public void setSize(long size) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    public void setName(String name) {
+        Assert.hasText(name, "File name is required");
```
Validation makes no sense according to declaration.

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    public Set<SourceFilePath> getSourceFilePaths() {
+        return sourceFilePaths;
```
Don't return the internal list, return a unmodifiable collection to avoid future problems.

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+            for (SourceFilePath sourceFilePath : sourceFilePaths) {
+                sourceFilePath.setFileMetadata(this);
+            }
+            this.sourceFilePaths = sourceFilePaths;
```
Don't store directly the collection, add all the items of the collection to your own. Initializate the sourceFilePaths in constructor.

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        return createdDate;
+    }
+
+    public void setCreatedDate(LocalDateTime createdDate) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        return lastModifiedDate;
+    }
+
+    public void setLastModifiedDate(LocalDateTime lastModifiedDate) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        return path;
+    }
+
+    public void setPath(String path) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        this.path = path;
+    }
+
+    public LocalDateTime getCreatedDate() {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        return createdDate;
+    }
+
+    public void setCreatedDate(LocalDateTime createdDate) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        this.createdDate = createdDate;
+    }
+
+    public LocalDateTime getLastModifiedDate() {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        return lastModifiedDate;
+    }
+
+    public void setLastModifiedDate(LocalDateTime lastModifiedDate) {
```
Not used method

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,107 @@
+        this.fileMetadata = fileMetadata;
+    }
+
+    public SourceFilePath() {
```
If not used, protected

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,157 @@
+        this.name = name;
+    }
+
+    public FileMetadata() {
```
If not used, protected

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/storage/FileSystemStorageService.java:
```diff
@@ -0,0 +1,82 @@
+
+    private final Path rootLocation;
+
+    @Autowired
```
No longer needed since Spring 4.3 -> [Spring 4.3 announcement](https://spring.io/blog/2016/03/04/core-container-refinements-in-spring-framework-4-3)

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/storage/StorageException.java:
```diff
@@ -0,0 +1,12 @@
+package uk.ac.ebi.ampt2d.storage;
+
+public class StorageException extends RuntimeException {
```
Why runtime exception?

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/controller/FileTrackingController.java:
```diff
@@ -0,0 +1,79 @@
+        File storedFile = convertMultipartToFile(multipartFile);
+
+        //TODO retrieve Type from request somehow
+        FileMetadata fileMetadata = new FileMetadata(storedFile, Type.VCF, multipartFile.getName());
```
Why don't asume other?

jorizci commented on src/main/java/uk/ac/ebi/ampt2d/Type.java:
```diff
@@ -0,0 +1,31 @@
+    GFF,
+    FASTA,
+    FASTQ,
+    OTHER
```
I think a generic BYTE type fits better than "OTHER"

jorizci committed:
File name changed to reflect repository entity name.

jorizci committed:
Type renamed to FileType.

jorizci committed:
Added binary type, added documentation

jorizci committed:
Changed constructor signature.

jorizci committed:
Changed StorageException from runtime to checked exception.

jorizci committed:
Moved Storage exceptions to storage/exceptions package

jorizci committed:
Modified FileTrackingController
-Removing unused getter/setter.
-Removed name.
-Modified type persistence to not null.

jorizci committed:
Refactor of FileSystemStorage

jorizci committed:
Refactor of system storage, tests working. Enabled Spring actuators

jorizci committed:
Added exception handling.

jorizci committed:
Removed unused imports. Cleaning of application.properties file.

jorizci committed:
Added license.

jorizci committed:
Added documentation

jorizci committed:
Added test for uploads

jorizci committed:
Added tests for metadata repository.

cyenyxe commented on pom.xml:
```diff
@@ -0,0 +1,127 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>uk.ac.ebi.eva</groupId>
```
The groupId in the accessioning service is `uk.ac.ebi.ampt2d`.

cyenyxe commented on pom.xml:
```diff
@@ -0,0 +1,127 @@
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+        </dependency>
```
We should prioritize PostgreSQL over MySQL.

cyenyxe commented on pom.xml:
```diff
@@ -0,0 +1,127 @@
+        </plugins>
+    </build>
+
+    <repositories>
```
Do we really need the `repositories` and `pluginRepositories` tags as they are now? The rest of our projects don't seem to use them.

jorizci replied:
They are their maven repositories for milestone / release candidate versions. We can remove it.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,121 @@
+    private LocalDateTime lastModifiedDate;
+
+    public FileMetadata(String hash, FileType fileType, long size) {
+        this();
```
The convention is usually the opposite, calling the constructor with more arguments from the one with less.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/FileMetadata.java:
```diff
@@ -0,0 +1,121 @@
+
+    private static final int MIN_FILE_HASH = 1;
+
+    private static final int MAX_FILE_HASH = 128;
```
`_LENGTH` suffix would clarify what these are.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/entities/SourceFilePath.java:
```diff
@@ -0,0 +1,84 @@
+public class SourceFilePath {
+
+    private static final int MIN_PATH = 1;
+    private static final int MAX_PATH = 1024;
```
`_LENGTH` suffix would clarify what these are.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/persistence/repository/FileMetadataRepository.java:
```diff
@@ -0,0 +1,37 @@
+ * Spring generated repository for accessing FileMetadata. This repository also includes the
+ * {@link RepositoryRestResource} that configures spring to generate automatically a rest controller for the repository.
+ */
+@RepositoryRestResource(collectionResourceRel = "files", path = FileMetadataRepository.REST_REPOSITORY_FILES)
```
Wouldn't it be simpler to make "/files" explicit as second argument?

jorizci replied:
This way it can be reused in the tests later.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/rest/FileTrackingController.java:
```diff
@@ -0,0 +1,52 @@
+    private FileTrackingService fileTrackingService;
+
+    @PostMapping(REST_UPLOAD)
+    public ResponseEntity<?> handleFileUpload(@RequestParam("file") MultipartFile multipartFile) throws
```
`upload` or `uploadFile`? `handle` sounds redundant.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/rest/FileTrackingController.java:
```diff
@@ -0,0 +1,52 @@
+    @Autowired
+    private FileTrackingService fileTrackingService;
+
+    @PostMapping(REST_UPLOAD)
```
As above, why not using the plain string?

jorizci replied:
As above, that way we can use it for the REST tests.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/rest/responses/ExceptionError.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Custom class to avoid leaking internal exception to the rest output message. {@link JsonView} annotation is added
+ * to the class members to avoid creating a converter to return it as a error response when a exception has occurred.
+ * Spring management of exceptions is outside of the control that provides automatic json converters.
```
Please care about capitalization, specially "REST output message" as the lower case could cause confusion with "rest of the output message".

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/rest/ExceptionHandling.java:
```diff
@@ -0,0 +1,45 @@
+
+    @ExceptionHandler({IOException.class, StorageException.class})
+    public ResponseEntity<ExceptionError> exceptionHandler(HttpServletRequest request, Exception exception) {
+        logger.error("Exception on rest call " + request.getRequestURI(), exception);
```
REST

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/services/FileTrackingService.java:
```diff
@@ -0,0 +1,61 @@
+    @Autowired
+    private FileSystemStorageService fileSystemStorageService;
+
+    public void addFileToTrackingService(InputStream inputStream) throws StorageException, IOException {
```
Redundant name `FileTrackingService.addFileToTrackingService` could be just `FileTrackingService.addFile`.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/services/FileTrackingService.java:
```diff
@@ -0,0 +1,61 @@
+
+            FileMetadata fileMetadata = new FileMetadata(fileHash, FileType.BINARY, fileSize);
+            fileMetadata.addSourceFilePaths(sourceFilePath);
+            fileMetadataRepository.save(fileMetadata);
```
Is there a way to wrap the filesystem and the database operation into a single transaction?

jorizci replied:
This is the only way, the java transactional system only works for databases. There is no simple way to include code for a broken transaction.

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/storage/FileSystemStorageService.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Override
+    public void remove(SourceFilePath sourceFilePath) {
```
`delete`

cyenyxe commented on src/main/java/uk/ac/ebi/ampt2d/storage/StorageProperties.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Folder location for storing files
+     */
+    private String location = "upload-dir";
```
Is this a default in case the value is not defined in the properties file?

cyenyxe commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,37 @@
+# Prefix of the application path (eg. http://localhost:8082/test/files). Will work with spring boot 1.5 and not with 2.0
+server.contextPath= /test
+
+spring.http.multipart.max-file-size=128KB
```
We are pretty much screwed if we only accept 128KB files :stuck_out_tongue: Can we set a more appropriate default?

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/configuration/RestTemplateConfiguration.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class RestTemplateConfiguration {
```
What is this needed for?

jorizci replied:
Is required to inject the TestRestTemplate tool.

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/persistence/repository/FileMetadataRepositoryTest.java:
```diff
@@ -0,0 +1,138 @@
+        fileMetadataRepository.save(fastq);
+    }
+
+    private void configureTemporalFilePaths() {
```
"temporary" instead of "temporal"

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/FileManagerApplicationTests.java:
```diff
@@ -0,0 +1,109 @@
+
+        ResponseEntity<Void> result = restTemplate.postForEntity(REST_UPLOAD, parts, Void.class);
+
+        Assert.assertEquals(HttpStatus.CREATED,result.getStatusCode());
```
This class uses `Assert.assertXXX` whereas others use `assertXXX` directly.

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/persistence/repository/FileMetadataRepositoryTest.java:
```diff
@@ -0,0 +1,138 @@
+    public void testLoadFiles() {
+        configureTemporalFilePaths();
+
+        List<FileMetadata> fileMetadatas = (ArrayList<FileMetadata>) fileMetadataRepository.findAll();
```
"metadatas" is not a correct plural noun.

jorizci committed:
Unified asserts used in test.

jorizci committed:
Modified temporal for temporary, removed unused imports.

jorizci committed:
Added documentation and license, removed unused imports

jorizci committed:
Added storage location to tmp, modified upload size limits.

jorizci committed:
Remove default value in storageProperties

jorizci committed:
Changed remove for delete.

jorizci committed:
changed addFileToTrackingService to addFile.

jorizci committed:
Modified rest to REST where applicable.

jorizci committed:
Changed method name from handleFileUpload to upload.

jorizci committed:
Changed constants to include _LENGTH where appropriate.

jorizci committed:
Modified pom to remove external repositories, mysql connector, modified
group id naming.

cyenyxe commented on src/test/java/uk/ac/ebi/ampt2d/FileManagerApplicationTests.java:
```diff
@@ -0,0 +1,98 @@
+        ResponseEntity<Void> result = restTemplate.postForEntity(REST_UPLOAD, parts, Void.class);
+        assertEquals(HttpStatus.CREATED,result.getStatusCode());
+        result = restTemplate.postForEntity(REST_UPLOAD, parts, Void.class);
+        assertEquals(HttpStatus.CREATED,result.getStatusCode());
```
Agreed offline to make the double upload and double register return the same HTTP code on the second request.

jorizci committed:
Modifie add file and upload behaviour.
-upload now returns a file metadata json response
-double post of the same file / hash now receives the same response as a doble upload.

jorizci committed:
Modified spring-boot version.

jorizci committed:
Added extra documentation to clarify some usage of Spring rest hateoas.

Commit:
More consistent naming in comments
